services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/model/BaseDto.java

```diff
@@ -0,0 +1,82 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.deviceregistry.mongodb.model;
+
+import java.time.Instant;
+import java.util.Objects;
+
+import org.eclipse.hono.annotation.HonoTimestamp;
+import org.eclipse.hono.util.RegistryManagementConstants;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * TODO.
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public abstract class BaseDto {
+
+    @JsonProperty(value = RegistryManagementConstants.FIELD_VERSION, required = true)
+    protected String version;
+    @JsonProperty(value = RegistryManagementConstants.FIELD_UPDATED_ON, required = true)
+    @HonoTimestamp
+    protected Instant updatedOn;
+
+    /**
+     * Default constructor for serialisation/deserialization.
+     */
+    public BaseDto() {
+        // Explicit default constructor.
+    }
+
+    /**
+     * Gets the version of the document.
+     * 
+     * @return The version of the document or {@code null} if not set.
+     */
+    public final String getVersion() {
```

Comment:
do we need an explicit version if we already have the *last updated* timestamp?

Reply:
The version is mapped to _Etags_ and also used to verify them.

Reply:
I know, that's why I am asking. We can simply use the *last updated* timestamp as the version ...

Reply:
I guess that is as unique as the UUID. In that case, we can use it.

Reply:
Neither the version nor the timestamp are nor need to be unique. Their only purpose is to serve as a reference for implementing optimistic locking for specific DTO instances.

Reply:
May be I should have added more info instead of just mentioning it to be unique. It need not be unique always and it could be reused. My point was in order to have a strong validator, the RFC spec suggests that it should be _unique across all versions of all representations associated with a particular resource over time_. My concern was, what if a resource has been updated more than one time within a very short span of time lets say a millisecond (eg. Automated scripts to update resources). I think this could be effectively handled by also making use of the nanoseconds as below.
`final String version = String.format("%d-%d", lastUpdatedOn.toEpochMilli(), lastUpdatedOn.getNano())`

Reply:
good point :-) no matter how fine the resolution of the timestamp used will be, we might still run into the problem of having two updates occurring at (almost) the same time ... based on that it is probably better to keep the strictly increasing sequence number as the version. Thanks for pointing that out :+1:

Reply:
@sophokles73 I have used now an increasing sequence number as the version and also implemented other comments. I have rebased with the master and pushed the changes now. Last 3 commits are the new ones for review. Would you mind taking a look again?